--- /tmp/orig-extract/g39ddc-1.11/driver/g39ddc.c	2018-06-06 07:24:59.000000000 -0400
+++ /home/brendan/winradio_devpack/g39ddc-1.11/driver/g39ddc.c	2026-06-20 15:54:22.288386992 -0400
@@ -520,7 +520,7 @@
 #else
 static void g39_wd_timer(struct timer_list *t)
 {
- struct g39ddc *dev=from_timer(dev,t,wd_timer);
+ struct g39ddc *dev=container_of(t,struct g39ddc,wd_timer);
 #endif
 
     queue_work(dev->workqueue,&dev->wd_work);
@@ -573,7 +573,7 @@
     g39_mutex_init(&dev->mutex);
     init_waitqueue_head(&dev->power_wait_queue);
 
-    dev->last_power_off_time=current_kernel_time();
+    ktime_get_real_ts64(&dev->last_power_off_time);
 
     for(i=0;i<G39_IO_STREAM_COUNT;i++)
     {
@@ -716,7 +716,7 @@
 static int g39_set_power(struct g39ddc *dev,uint8_t power,uint8_t lock)
 {
  int status;
- struct timespec time;
+ struct timespec64 time;
  int i;
 
     power=power?1:0;
@@ -737,7 +737,7 @@
             dev->state.suspended=0;
             dev->state.temperature=0;
 
-            time=current_kernel_time();
+            ktime_get_real_ts64(&time);
             time.tv_nsec-=500000000;
 
             if(time.tv_nsec<0)
@@ -747,7 +747,7 @@
             }
 
             if((dev->last_power_off_time.tv_sec==0 && dev->last_power_off_time.tv_nsec==0) ||
-               timespec_compare(&time,&dev->last_power_off_time)<0)
+               timespec64_compare(&time,&dev->last_power_off_time)<0)
             {
                 msleep(500);
             }
@@ -779,7 +779,7 @@
             {
                 dev->state.power=0;
                 dev->state.fpga=0;
-                dev->last_power_off_time=current_kernel_time();
+                ktime_get_real_ts64(&dev->last_power_off_time);
 
                 wake_up_all(&dev->power_wait_queue);
             }
@@ -803,7 +803,7 @@
         return -EINVAL;
     }
 
-    if(unlikely(!access_ok(VERIFY_READ,user_buffer,size)))
+    if(unlikely(!access_ok(user_buffer,size)))
     {
         return -EFAULT;
     }
@@ -949,7 +949,7 @@
         return -EINVAL;
     }
 
-    if(unlikely(!access_ok(VERIFY_WRITE,(void __user*)arg,sizeof(void*))))
+    if(unlikely(!access_ok((void __user*)arg,sizeof(void*))))
     {
         return -EFAULT;
     }
@@ -971,7 +971,7 @@
  uintptr_t user_buffer;
  int status;
 
-    if(unlikely(!access_ok(VERIFY_WRITE,(void __user*)compat_ptr(arg),sizeof(compat_uptr_t))))
+    if(unlikely(!access_ok((void __user*)compat_ptr(arg),sizeof(compat_uptr_t))))
     {
         return -EFAULT;
     }
@@ -1554,14 +1554,14 @@
  int status;
  void *buffer;
 
-    if(unlikely(transferred!=NULL && !access_ok(transferred,VERIFY_WRITE,sizeof(uint16_t))))
+    if(unlikely(transferred!=NULL && !access_ok(transferred,sizeof(uint16_t))))
     {
         return -EFAULT;
     }
 
     if(buffer_size!=0)
     {
-        if(!access_ok(user_buffer,VERIFY_WRITE,buffer_size))
+        if(!access_ok(user_buffer,buffer_size))
         {
             return -EFAULT;
         }
@@ -1623,7 +1623,7 @@
  int status;
  void *buffer;
 
-    if(unlikely(transferred!=NULL && !access_ok(transferred,VERIFY_WRITE,sizeof(uint16_t))))
+    if(unlikely(transferred!=NULL && !access_ok(transferred,sizeof(uint16_t))))
     {
         return -EFAULT;
     }
@@ -1681,14 +1681,14 @@
  int len;
  void *buffer;
 
-    if(unlikely(transferred!=NULL && !access_ok(transferred,VERIFY_WRITE,sizeof(uint16_t))))
+    if(unlikely(transferred!=NULL && !access_ok(transferred,sizeof(uint16_t))))
     {
         return -EFAULT;
     }
 
     if(buffer_size!=0)
     {
-        if(!access_ok(user_buffer,VERIFY_WRITE,buffer_size))
+        if(!access_ok(user_buffer,buffer_size))
         {
             return -EFAULT;
         }
@@ -1755,7 +1755,7 @@
  int len;
  void *buffer;
 
-    if(unlikely(transferred!=NULL && !access_ok(transferred,VERIFY_WRITE,sizeof(uint16_t))))
+    if(unlikely(transferred!=NULL && !access_ok(transferred,sizeof(uint16_t))))
     {
         return -EFAULT;
     }
